pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -91,19 +91,19 @@ public class ModelInterpreter implements Serializable {
                                                     // environment is populated before we evaluate any when condition,
                                                     // and so that we don't go into a per-stage agent if the when condition
                                                     // isn't satisfied.
-                                                    withEnvBlock(thisStage.getEnvVars(root, script)) {
-                                                        if (evaluateWhen(thisStage.when)) {
-                                                            inDeclarativeAgent(thisStage, root, thisStage.agent) {
-                                                                withCredentialsBlock(thisStage.getEnvCredentials()) {
+                                                    withCredentialsBlock(thisStage.getEnvCredentials()) {
+                                                        withEnvBlock(thisStage.getEnvVars(root, script)) {
+                                                            if (evaluateWhen(thisStage.when)) {
+                                                                inDeclarativeAgent(thisStage, root, thisStage.agent) {
```

Comment:
:bug: `withCredentials` step requires a workspace so this will only work when there is a global agent allocated.